src/verifier/step3/Step3Logic.sol

```diff
@@ -0,0 +1,359 @@
+
+        SpongeOpLib.SpongeOp memory absorb = SpongeOpLib.SpongeOp(SpongeOpLib.SpongeOpType.Absorb, uint32(counter));
+        SpongeOpLib.SpongeOp memory squeeze = SpongeOpLib.SpongeOp(SpongeOpLib.SpongeOpType.Squeeze, 1);
+        SpongeOpLib.SpongeOp[] memory pattern = new SpongeOpLib.SpongeOp[](2);
+        pattern[0] = absorb;
+        pattern[1] = squeeze;
+        IOPatternLib.IOPattern memory p = IOPatternLib.IOPattern(pattern);
+        
+        NovaSpongePallasLib.SpongeU24Pallas memory sponge = NovaSpongePallasLib.start(p, 0);
+
+        sponge = NovaSpongePallasLib.absorb(sponge, elementsToHash);
+
+        (, uint256[] memory output) = NovaSpongePallasLib.squeeze(sponge, 1);
+        sponge = NovaSpongePallasLib.finishNoFinalIOCounterCheck(sponge);
+
+        // uint256 r = output[0] & 0x00000000000000000000000000000000ffffffffffffffffffffffffffffffff;
+
+        RelaxedR1CSInstance memory result = foldInstance(U1, U2, comm_T, output[0] & 0x00000000000000000000000000000000ffffffffffffffffffffffffffffffff);
+
+        return result;
+    }
+
+    function foldInstance(
+        RelaxedR1CSInstance calldata U1,
+        R1CSInstance calldata u2,
+        Pallas.PallasAffinePoint memory comm_T,
+        uint256 r
+    ) public view returns (RelaxedR1CSInstance memory) {
+        // uint256[] memory x1 = U1.X;
+        // Pallas.PallasAffinePoint memory comm_W_1 = U1.comm_W;
+        // Pallas.PallasAffinePoint memory comm_E_1 = U1.comm_E;
+        // uint256 u1 = U1.u;
+
+        uint256[] memory x2 = u2.X;
+        Pallas.PallasAffinePoint memory comm_W_2 = u2.comm_W;
+
+        require(U1.X.length == x2.length, "Witness vectors do not match length");
+
+        uint256[] memory X = new uint256[](U1.X.length);
+
+        for (uint256 i = 0; i < x2.length; i++) {
+            X[i] = addmod(U1.X[i], mulmod(r, x2[i], Pallas.R_MOD), Pallas.R_MOD);
+        }
+
+        // Pallas.PallasAffinePoint memory comm_W = Pallas.add(comm_W_1, Pallas.scalarMul(comm_W_2, r));
+
+        // Pallas.PallasAffinePoint memory comm_E = Pallas.add(comm_E_1, Pallas.scalarMul(comm_T, r));
+
+        // uint256 u = u1 + r;
+
+        return RelaxedR1CSInstance(Pallas.add(U1.comm_W, Pallas.scalarMul(comm_W_2, r)) , Pallas.add(U1.comm_E, Pallas.scalarMul(comm_T, r)) , X, addmod(U1.u, r, Pallas.P_MOD));
+    }
+}
+
+// TODO: copy the modifications from the Pallas library to the Vesta library
+library NIFSVesta {
+    uint256 constant private MOD = Vesta.P_MOD;
+    uint256 constant private NUM_FE_FOR_RO = 24;
+
+    struct NIFS {
+        uint8[32] compressed_comm_T;
+    }
+
+    struct R1CSInstance {
+        Vesta.VestaAffinePoint comm_W;
+        uint256[] X;
+    }
+
+    struct RelaxedR1CSInstance {
+        Vesta.VestaAffinePoint comm_W;
+        Vesta.VestaAffinePoint comm_E;
+        uint256[] X;
+        uint256 u;
+    }
+
+    function verify(
+        NIFS memory nifs,
+        uint256 pp_digest,
+        RelaxedR1CSInstance calldata U1,
```

Comment:
Could you please explain, why did you decide to use `calldata`, not `memory` here?

Reply:
My understanding is that `calldata` has a lower gas cost than memory, so if possible one should prefer it over `memory`. The downside being that mutability is not allowed.